go/vt/automation/scheduler.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2015, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+/*
+Package automation contains code to execute high-level cluster operations
+(e.g. resharding) as a series of low-level operations
+(e.g. vtctl, shell commands, ...).
+*/
+package automation
+
+import (
+	"fmt"
+	"sync"
+
+	log "github.com/golang/glog"
+	pb "github.com/youtube/vitess/go/vt/proto/automation"
+	"golang.org/x/net/context"
+)
+
+type schedulerState int32
+
+const (
+	stateNotRunning schedulerState = iota
+	stateRunning
+	stateShuttingDown
+	stateShutdown
+)
+
+// Scheduler executes automation tasks and maintains the execution state.
+type Scheduler struct {
+	idGenerator IDGenerator
+
+	mu sync.Mutex
+	// Guarded by "mu".
+	registeredClusterOperations map[string]bool
+	// Guarded by "mu".
+	toBeScheduledClusterOperations chan *ClusterOperationInstance
+	// Guarded by "mu".
+	state schedulerState
+
+	pendingOpsWg *sync.WaitGroup
+
+	muOpList sync.Mutex
+	// Guarded by "muOpList".
+	activeClusterOperations map[string]*ClusterOperationInstance
+	// Guarded by "muOpList".
+	finishedClusterOperations map[string]*ClusterOperationInstance
+}
+
+// NewScheduler creates a new instance.
+func NewScheduler() (*Scheduler, error) {
+	defaultClusterOperations := map[string]bool{
+		"ReshardingTask": true,
+		// Testing only cluster operations.
+		"TestingEchoTask":     true,
+		"TestingEmitEchoTask": true,
+	}
+
+	s := &Scheduler{
+		registeredClusterOperations:    defaultClusterOperations,
+		idGenerator:                    IDGenerator{},
+		toBeScheduledClusterOperations: make(chan *ClusterOperationInstance, 10),
+		state:                     stateNotRunning,
+		pendingOpsWg:              &sync.WaitGroup{},
+		activeClusterOperations:   make(map[string]*ClusterOperationInstance),
+		finishedClusterOperations: make(map[string]*ClusterOperationInstance),
+	}
+
+	return s, nil
+}
+
+// Run processes queued cluster operations.
+func (s *Scheduler) Run() {
+	s.mu.Lock()
+	s.state = stateRunning
+	s.mu.Unlock()
+
+	s.pendingOpsWg.Add(1)
```

Comment:
Oh, interesting, this is a good point. I think you should make it clear in the comments that there is a very good reason for doing what you're doing. Something like:

// Use a wait-group instead of just a done channel, because we want
// to be able to shut down the scheduler even if Run() was never executed.